File: Foundation.Data.Doublets.Cli.Tests/UnicodeStringStorageTests.cs
using Platform.Data;
using Platform.Data.Doublets;
using Platform.Data.Doublets.Memory.United.Generic;
using Platform.Memory;

namespace Foundation.Data.Doublets.Cli.Tests
{
    public class UnicodeStringStorageTests
    {
        [Fact]
        public void CreateAndRetrieveEmptyStringTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // Act
                var emptyStringLink = storage.CreateString("");
                var retrievedString = storage.GetString(emptyStringLink);

                // Assert
                Assert.Equal("", retrievedString);
            });
        }

        [Fact]
        public void CreateAndRetrieveSimpleStringTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // Act
                var helloLink = storage.CreateString("Hello");
                var retrievedHello = storage.GetString(helloLink);

                // Assert
                Assert.Equal("Hello", retrievedHello);
            });
        }

        [Fact]
        public void CreateAndRetrieveMultipleStringsTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // Act
                var linkOne = storage.CreateString("First");
                var linkTwo = storage.CreateString("Second");
                var retrievedOne = storage.GetString(linkOne);
                var retrievedTwo = storage.GetString(linkTwo);

                // Assert
                Assert.Equal("First", retrievedOne);
                Assert.Equal("Second", retrievedTwo);
            });
        }

        [Fact]
        public void CreateAndRetrieveUnicodeStringTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // A string with some unicode characters
                var unicodeContent = "Hello, 世界! Привет, мир!";

                // Act
                var unicodeLink = storage.CreateString(unicodeContent);
                var retrieved = storage.GetString(unicodeLink);

                // Assert
                Assert.Equal(unicodeContent, retrieved);
            });
        }

        [Fact]
        public void RetrieveTypeByNameTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // Act
                var retrievedType = storage.NamedLinks.GetByName("Type");

                // Assert
                Assert.Equal(1u, retrievedType);
            });
        }

        [Theory]
        [InlineData("Type")]
        [InlineData("UnicodeSymbol")]
        [InlineData("UnicodeSequence")]
        [InlineData("String")]
        [InlineData("EmptyString")]
        [InlineData("Name")]
        public void CreateAndRetrieveMultipleStringTypesTest(string typeName)
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // Act
                var retrievedType = storage.NamedLinks.GetByName(typeName);

                // Assert
                Assert.Equal(typeName, storage.NamedLinks.GetName(retrievedType));
            });
        }

        [Fact]
        public void CreateAndRetriveUserDefinedTypeTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                // Act
                var userType = storage.GetOrCreateType("UserType");
                var retrievedType = storage.NamedLinks.GetByName("UserType");

                // Assert
                Assert.Equal(userType, retrievedType);
            });
        }

        [Fact]
        public void NameExternalReferenceTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);

                Hybrid<uint> hybrid = new Hybrid<uint>(1, isExternal: true);

                // Act
                storage.NamedLinks.SetName(hybrid, "MyExternalReference");
                var retrievedName = storage.NamedLinks.GetName(hybrid);
                var retrievedIndex = storage.NamedLinks.GetByName("MyExternalReference");

                // Assert
                Assert.Equal(4294967295ul, (ulong)hybrid);
                Assert.Equal(4294967295, retrievedIndex);
                Assert.Equal(1ul, (ulong)hybrid.AbsoluteValue);
                Assert.True(hybrid.IsExternal);
                Assert.Equal("MyExternalReference", retrievedName);
            });
        }

        [Fact]
        public void NameIsRemovedWhenLinkIsDeletedTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);
                var link = links.Create(new uint[] { 0, 0 });
                storage.NamedLinks.SetName(link, "TestName");
                Assert.Equal("TestName", storage.NamedLinks.GetName(link));
                Assert.Equal(link, storage.NamedLinks.GetByName("TestName"));

                // Act: delete the link and its name
                links.Delete(link);
                storage.NamedLinks.RemoveName(link);

                // Assert: name is removed
                Assert.Equal(links.Constants.Null, storage.NamedLinks.GetByName("TestName"));
                Assert.Null(storage.NamedLinks.GetName(link));
            });
        }

        [Fact]
        public void DeletingNonNamedLinkDoesNotAffectOtherNamesTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);
                var namedLink = links.Create(new uint[] { 0, 0 });
                storage.NamedLinks.SetName(namedLink, "Named");
                var unnamedLink = links.Create(new uint[] { 0, 0 });
                Assert.Equal("Named", storage.NamedLinks.GetName(namedLink));

                // Act: delete the unnamed link
                links.Delete(unnamedLink);

                // Assert: named link and its name remain
                Assert.Equal(namedLink, storage.NamedLinks.GetByName("Named"));
                Assert.Equal("Named", storage.NamedLinks.GetName(namedLink));
            });
        }

        [Fact]
        public void NameIsRemovedWhenExternalReferenceIsDeletedTest()
        {
            RunTestWithLinks(links =>
            {
                // Arrange
                var storage = new UnicodeStringStorage<uint>(links);
                var externalRef = 123u;
                storage.NamedLinks.SetNameForExternalReference(externalRef, "ExternalName");
                Assert.Equal("ExternalName", storage.NamedLinks.GetNameByExternalReference(externalRef));
                Assert.Equal(externalRef, storage.NamedLinks.GetExternalReferenceByName("ExternalName"));

                // Act: remove the name for the external reference (do not call links.Delete for external references)
                storage.NamedLinks.RemoveNameByExternalReference(externalRef);

                // Assert: name is removed
                Assert.Equal(links.Constants.Null, storage.NamedLinks.GetExternalReferenceByName("ExternalName"));
                Assert.Null(storage.NamedLinks.GetNameByExternalReference(externalRef));
            });
        }

        // Helper method to create a test environment with a temporary file
        private static void RunTestWithLinks(Action<ILinks<uint>> testAction)
        {
            var tempDbFile = Path.GetTempFileName();
            try
            {
                var constants = new LinksConstants<uint>(enableExternalReferencesSupport: true);
                var memory = new FileMappedResizableDirectMemory(tempDbFile, UnitedMemoryLinks<uint>.DefaultLinksSizeStep);
                using var links = new UnitedMemoryLinks<uint>(memory, UnitedMemoryLinks<uint>.DefaultLinksSizeStep, constants, Platform.Data.Doublets.Memory.IndexTreeType.Default);
                var decoratedLinks = links.DecorateWithAutomaticUniquenessAndUsagesResolution();
                testAction(decoratedLinks);
            }
            finally
            {
                File.Delete(tempDbFile);
            }
        }
    }
}

File: rust/tests/unicode_string_storage_tests.rs
use anyhow::Result;
use link_cli::{external_reference, LinkStorage, NamedLinks, UnicodeStringStorage};
use tempfile::NamedTempFile;

fn with_storage(test: impl FnOnce(&mut UnicodeStringStorage<'_>) -> Result<()>) -> Result<()> {
    let temp_file = NamedTempFile::new()?;
    let db_path = temp_file.path().to_str().unwrap();
    let mut links = LinkStorage::new(db_path, false)?;
    let mut storage = UnicodeStringStorage::new(&mut links)?;
    test(&mut storage)
}

#[test]
fn create_and_retrieve_empty_string() -> Result<()> {
    with_storage(|storage| {
        let empty = storage.create_string("")?;
        assert_eq!("", storage.get_string(empty)?);
        Ok(())
    })
}

#[test]
fn create_and_retrieve_simple_string() -> Result<()> {
    with_storage(|storage| {
        let hello = storage.create_string("Hello")?;
        assert_eq!("Hello", storage.get_string(hello)?);
        Ok(())
    })
}

#[test]
fn create_and_retrieve_multiple_strings() -> Result<()> {
    with_storage(|storage| {
        let first = storage.create_string("First")?;
        let second = storage.create_string("Second")?;

        assert_eq!("First", storage.get_string(first)?);
        assert_eq!("Second", storage.get_string(second)?);
        Ok(())
    })
}

#[test]
fn create_and_retrieve_unicode_string_as_utf16_sequence() -> Result<()> {
    with_storage(|storage| {
        let content = "Hello, 世界! Привет, мир! 😀";
        let link = storage.create_string(content)?;

        assert_eq!(content, storage.get_string(link)?);
        assert!(storage.unicode_sequence_code_units(link)?.len() > content.chars().count());
        Ok(())
    })
}

#[test]
fn pinned_types_are_created_and_named() -> Result<()> {
    with_storage(|storage| {
        assert_eq!(1, storage.type_type());
        assert_eq!(2, storage.unicode_symbol_type());
        assert_eq!(3, storage.unicode_sequence_type());
        assert_eq!(4, storage.string_type());
        assert_eq!(5, storage.empty_string_type());
        assert_eq!(6, storage.name_type());

        for (id, name) in [
            (storage.type_type(), "Type"),
            (storage.unicode_symbol_type(), "UnicodeSymbol"),
            (storage.unicode_sequence_type(), "UnicodeSequence"),
            (storage.string_type(), "String"),
            (storage.empty_string_type(), "EmptyString"),
            (storage.name_type(), "Name"),
        ] {
            assert_eq!(Some(id), storage.get_by_name(name)?);
            assert_eq!(Some(name.to_string()), storage.get_name(id)?);
        }

        Ok(())
    })
}

#[test]
fn create_and_retrieve_user_defined_type() -> Result<()> {
    with_storage(|storage| {
        let user_type = storage.get_or_create_type("UserType")?;
        assert_eq!(Some(user_type), storage.get_by_name("UserType")?);
        assert_eq!(Some("UserType".to_string()), storage.get_name(user_type)?);
        Ok(())
    })
}

#[test]
fn name_external_reference_matches_csharp_hybrid_encoding() -> Result<()> {
    with_storage(|storage| {
        assert_eq!(u32::MAX, external_reference(1));

        storage.set_name_for_external_reference(1, "MyExternalReference")?;

        assert_eq!(
            Some("MyExternalReference".to_string()),
            storage.get_name_by_external_reference(1)?
        );
        assert_eq!(
            Some(1),
            storage.get_external_reference_by_name("MyExternalReference")?
        );
        assert_eq!(
            Some("MyExternalReference".to_string()),
            storage.get_name(external_reference(1))?
        );

        Ok(())
    })
}

#[test]
fn name_is_removed_when_link_is_deleted() -> Result<()> {
    with_storage(|storage| {
        let link = storage.links_mut().create(0, 0);
        storage.set_name(link, "TestName")?;
        assert_eq!(Some("TestName".to_string()), storage.get_name(link)?);
        assert_eq!(Some(link), storage.get_by_name("TestName")?);

        storage.links_mut().delete(link)?;
        storage.remove_name(link)?;

        assert_eq!(None, storage.get_by_name("TestName")?);
        assert_eq!(None, storage.get_name(link)?);
        Ok(())
    })
}

#[test]
fn deleting_non_named_link_does_not_affect_other_names() -> Result<()> {
    with_storage(|storage| {
        let named_link = storage.links_mut().create(0, 0);
        storage.set_name(named_link, "Named")?;
        let unnamed_link = storage.links_mut().create(0, 0);

        storage.links_mut().delete(unnamed_link)?;

        assert_eq!(Some(named_link), storage.get_by_name("Named")?);
        assert_eq!(Some("Named".to_string()), storage.get_name(named_link)?);
        Ok(())
    })
}

#[test]
fn name_is_removed_when_external_reference_is_deleted() -> Result<()> {
    with_storage(|storage| {
        storage.set_name_for_external_reference(123, "ExternalName")?;
        assert_eq!(
            Some("ExternalName".to_string()),
            storage.get_name_by_external_reference(123)?
        );
        assert_eq!(
            Some(123),
            storage.get_external_reference_by_name("ExternalName")?
        );

        storage.remove_name_by_external_reference(123)?;

        assert_eq!(
            None,
            storage.get_external_reference_by_name("ExternalName")?
        );
        assert_eq!(None, storage.get_name_by_external_reference(123)?);
        Ok(())
    })
}

#[test]
fn named_links_facade_matches_csharp_named_links_role() -> Result<()> {
    let temp_file = NamedTempFile::new()?;
    let db_path = temp_file.path().to_str().unwrap();
    let mut links = LinkStorage::new(db_path, false)?;
    let mut named_links = NamedLinks::new(&mut links)?;

    let link = named_links.unicode_storage_mut().links_mut().create(0, 0);
    named_links.set_name(link, "FacadeName")?;

    assert_eq!(Some(link), named_links.get_by_name("FacadeName")?);
    assert_eq!(Some("FacadeName".to_string()), named_links.get_name(link)?);
    Ok(())
}
